### PR TITLE
UCT/CUDA/GTEST: Enable P2P RMA testing with cuda memory

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5666/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5668/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5669/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5670/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5671/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5680/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5682/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5690/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5693/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
@yosefe @alex-mikheev  please review

File: test/gtest/uct/uct_test.cc [hunk 1]
```diff
@@ -655,28 +697,56 @@ uct_test::mapped_buffer::mapped_buffer(size_t size, uint64_t seed,
...
+    if (mem_type == UCT_MD_MEM_TYPE_HOST) {
+        while ((char*)(ptr + 1) <= end) {
+            *ptr = seed;
+            seed = pat(seed);
+            ++ptr;
+        }
+        memcpy(ptr, &seed, end - (char*)ptr);
+    } else if (mem_type == UCT_MD_MEM_TYPE_CUDA) {
+#if HAVE_CUDA
+        void *temp;
```
Comment (alex-mikheev) [inline test/gtest/uct/uct_test.cc]: may be it is a good idea to move cuda handling code to separate functions.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5737/ for details (Mellanox internal link).

File: test/gtest/uct/uct_test.cc (hunk 1 above)
Comment (bureddy) [reply test/gtest/uct/uct_test.cc]: moved cuda pattern fill/check to separate funcs

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5740/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5751/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5752/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5753/ for details (Mellanox internal link).

File: test/gtest/uct/test_p2p_rma.cc [hunk 2]
```diff
@@ -62,18 +62,25 @@ ucs_status_t uct_p2p_rma_test::get_zcopy(uct_ep_h ep, const mapped_buffer &sendb
...
-    blocking_send(send, sender_ep(), sendbuf, recvbuf, true);
-    if (direction == DIRECTION_SEND_TO_RECV) {
-        sendbuf.pattern_fill(SEED3);
-        wait_for_remote();
-        recvbuf.pattern_check(SEED1);
-    } else if (direction == DIRECTION_RECV_TO_SEND) {
-        recvbuf.pattern_fill(SEED3);
-        sendbuf.pattern_check(SEED2);
-        wait_for_remote();
+    for (int mem_type = 0; mem_type < UCT_MD_MEM_TYPE_LAST; mem_type++) {
```
Comment (yosefe) [inline test/gtest/uct/test_p2p_rma.cc]: can we make this iteration inside uct_p2p_test::test_xfer_multi instead?
e.g
1. rename test_xfer_multi() to test_xfer_multi_mem_type(..., mem_type)
2. test_xfer_multi() will iterate over memtype and call test_xfer_multi_mem_type()
3. test_xfer_multi() will print to the output ("ms") the tested memory type, in the beginning of the line, so we could see which memtypes are being tested

File: test/gtest/uct/uct_test.cc [hunk 3]
```diff
@@ -334,17 +334,42 @@ uct_test::entity::entity(const resource& resource, uct_iface_config_t *iface_con
...
 
     if (md_attr().cap.flags & (UCT_MD_FLAG_ALLOC|UCT_MD_FLAG_REG)) {
-        status = uct_iface_mem_alloc(m_iface, length, UCT_MD_MEM_ACCESS_ALL,
-                                     alloc_name, mem);
-        ASSERT_UCS_OK(status);
+        if (mem_type == UCT_MD_MEM_TYPE_HOST) {
+            status = uct_iface_mem_alloc(m_iface, length, UCT_MD_MEM_ACCESS_ALL,
+                                         alloc_name, mem);
+            ASSERT_UCS_OK(status);
+        } else if(mem_type == UCT_MD_MEM_TYPE_CUDA) {
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: space after if

File: test/gtest/uct/uct_test.cc [hunk 4]
```diff
@@ -334,17 +334,42 @@ uct_test::entity::entity(const resource& resource, uct_iface_config_t *iface_con
...
+#if HAVE_CUDA
+            cudaError_t cerr;
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: maybe extract the code which allocates cuda mem and registers it on a memory domain to a helper function?

File: test/gtest/uct/uct_test.cc [hunk 5]
```diff
@@ -378,15 +403,30 @@ void uct_test::entity::mem_alloc(size_t length, uct_allocated_memory_t *mem,
...
-        uct_iface_mem_free(mem);
+    if (mem_type == UCT_MD_MEM_TYPE_HOST) {
+        if (mem->method != UCT_ALLOC_METHOD_LAST) {
+            uct_iface_mem_free(mem);
+        }
+    } else if(mem_type == UCT_MD_MEM_TYPE_CUDA) {
+#if HAVE_CUDA
+        cudaError_t cerr;
+
+        if (md_attr().cap.reg_mem_types & UCS_BIT(mem_type)) {
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: i would check if memh!=UCT_MEM_HANDLE_NULL here

File: test/gtest/uct/test_p2p_rma.cc (hunk 2 above)
Comment (bureddy) [reply test/gtest/uct/test_p2p_rma.cc]: I wanted to restrict this to rma, if I move iteration to uct_p2p_test::test_xfer_multi and pass mem_type to  subclasse (rma,am ..etc) test_xfer, then I have to ignore mem_type in uct_p2p_am_test.  If I do that same test will run twice with host memory because of mem type iteration.
For example  IB TL support cuda memory but not support short, bcopy..
is there anyway to handle this scenario?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5763/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5764/ for details (Mellanox internal link).

File: test/gtest/uct/test_p2p_rma.cc (hunk 2 above)
Comment (bureddy) [reply test/gtest/uct/test_p2p_rma.cc]: pushed a solution. please review.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5765/ for details (Mellanox internal link).

File: test/gtest/uct/test_p2p_am.cc [hunk 6]
```diff
@@ -204,13 +204,25 @@ class uct_p2p_am_test : public uct_p2p_test
         }
     }
 
-    virtual void test_xfer(send_func_t send, size_t length, direction_t direction) {
+    virtual int is_op_supported_with_mem_type(send_func_t send,
+                                              uct_memory_type_t mem_type) {
+        if (mem_type == UCT_MD_MEM_TYPE_CUDA) {
+            if ((send == &uct_p2p_am_test::am_short) ||
```
Comment (yosefe) [inline test/gtest/uct/test_p2p_am.cc]: why is that needed? AFAIK cuda TLs do not support AM anyway?

File: test/gtest/uct/uct_p2p_test.cc [hunk 7]
```diff
@@ -15,6 +15,8 @@ int             uct_p2p_test::log_data_count = 0;
 ucs_log_level_t uct_p2p_test::orig_log_level;
 
 
+std::string const uct_p2p_test::mem_types[] = {"host", "cuda"};
```
Comment (yosefe) [inline test/gtest/uct/uct_p2p_test.cc]: maybe we need to add this to UCT base? since the enum is defined in UCT
eg `const char *uct_mem_type_names[]`

File: test/gtest/uct/test_p2p_rma.cc (hunk 2 above)
Comment (yosefe) [reply test/gtest/uct/test_p2p_rma.cc]: but we do want to test IB's `am_zcopy` with cuda memory, right?

File: test/gtest/uct/test_p2p_am.cc (hunk 6 above)
Comment (yosefe) [reply test/gtest/uct/test_p2p_am.cc]: ok, i got it now. so maybe, instead of `direction`, pass down `flags`?
and the test will run cuda mem if both sides are zcopy
later on, we will add more flags which would allow also testing am_zcopy
```
TEST_UCT_FLAG_DIR_SEND_TO_RECV = UCS_BIT(0),
TEST_UCT_FLAG_SEND_ZCOPY = UCS_BIT(1),
TEST_UCT_FLAG_RECV_ZCOPY = UCS_BIT(2),
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5767/ for details (Mellanox internal link).

File: test/gtest/uct/test_p2p_am.cc (hunk 6 above)
Comment (bureddy) [reply test/gtest/uct/test_p2p_am.cc]: Fixed. please review
